presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1121,10 +1116,161 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
+                            column.getType(),
+                            false,
+                            Optional.of(name),
+                            Optional.of(column.getName()),
+                            false))
+                    .collect(toImmutableList());
+
+            analysis.addRelationCoercion(table, outputFields.stream().map(Field::getType).toArray(Type[]::new));
+
+            return createAndAssignScope(table, scope, outputFields);
+        }
+
+        private Scope processMaterializedView(
+                Table materializedView,
+                QualifiedObjectName materializedViewObjectName,
+                Optional<Scope> scope,
+                QualifiedObjectName materializedViewQualifiedObjectName,
+                ConnectorMaterializedViewDefinition materializedViewDefinition,
+                Statement statement,
+                MaterializedViewStatus materializedViewStatus)
+        {
+            validateMaterialziedViewQueryPlan(statement);
+            String newSql = getMaterializedViewSQL((Query) statement, materializedViewDefinition, materializedViewStatus);
+
+            Query query = (Query) sqlParser.createStatement(newSql, createParsingOptions(session, warningCollector));
+            analysis.registerNamedQuery(materializedView, query);
+
+            analysis.registerMaterializedViewForAnalysis(materializedView);
+            RelationType relationType = analyzeView(
+                    query,
+                    materializedViewQualifiedObjectName,
+                    Optional.of(materializedViewObjectName.getCatalogName()),
+                    Optional.of(materializedViewDefinition.getSchema()),
+                    materializedViewDefinition.getOwner(),
+                    materializedView);
+            analysis.unregisterMaterializedViewForAnalysis(materializedView);
+            return createAndAssignScope(materializedView, scope, relationType);
+        }
+
+        private String getMaterializedViewSQL(
+                Query statement,
+                ConnectorMaterializedViewDefinition connectorMaterializedViewDefinition,
+                MaterializedViewStatus materializedViewStatus)
+        {
+            String materializedViewCreateSql = connectorMaterializedViewDefinition.getOriginalSql();
+            if (materializedViewStatus.isNotMaterialized()) {
+                return materializedViewCreateSql;
+            }
+
+            checkState(materializedViewStatus.isPartiallyMaterialized(), "materialized view status is " + materializedViewStatus);
+
+            Statement createSqlStatement = sqlParser.createStatement(materializedViewCreateSql, createParsingOptions(session, warningCollector));
+            QuerySpecification createSqlQuerySpecification = (QuerySpecification) ((Query) createSqlStatement).getQueryBody();
+
+            Map<SchemaTableName, Expression> partitionPredicates = generatePartitionPredicate(materializedViewStatus.getPartitionsFromBaseTables());
```

Comment:
Add a major TODO here: we will need to add the predicates for materialized views as well. It is in general not an issue but could happen when we are in the process of this function and a new partition lands.